core/domain/skill_services.py

```diff
@@ -230,6 +235,251 @@ def get_all_skill_summaries():
+                batch. If False, there are no further results after this batch.
+    """
+    # The fetched skills will be filtered afterwards and filtering may result
+    # in having less number of skills than requested. Hence, fetching twice
+    # the number of requested skills will help reduce the number of datastore
+    # calls.
+    skill_summaries_models, new_urlsafe_start_cursor, more = (
+        skill_models.SkillSummaryModel.fetch_page(
+            2 * num_skills_to_fetch, urlsafe_start_cursor, sort_by))
+
+    skill_summaries = [
+        get_skill_summary_from_model(summary)
+        for summary in skill_summaries_models]
+    return skill_summaries, new_urlsafe_start_cursor, more
+
+
+def get_filtered_skill_summaries(
+        num_skills_to_fetch, status, classroom_name, keywords,
+        sort_by, urlsafe_start_cursor):
+    """Returns all the skill summary dicts after filtering.
+
+    Args:
+        num_skills_to_fetch: int. Number of skills to fetch.
+        status: str. The status of the skill.
+        classroom_name: str. The classroom_name of the topic to which the skill
+            is assigned to.
+        keywords: list(str). The keywords to look for
+            in the skill description.
+        sort_by: str. A string indicating how to sort the result.
+        urlsafe_start_cursor: str or None. The cursor to the next page.
+
+    Returns:
+        3-tuple(augmented_skill_summaries, new_urlsafe_start_cursor, more).
+        where:
+            augmented_skill_summaries: list(AugmentedSkillSummary). The list of
+                augmented skill summaries. The number of returned skills might
+                include more than the requested number. Hence, the cursor
+                returned will represent the point to which those results were
+                fetched (and not the "num_skills_to_fetch" point).
+            new_urlsafe_start_cursor: str or None. A query cursor pointing to
+                the next batch of results. If there are no more results, this
+                might be None.
+            more: bool. If True, there are (probably) more results after this
+                batch. If False, there are no further results after this batch.
+    """
+    augmented_skill_summaries = []
+
+    augmented_skill_summaries_batch, new_urlsafe_start_cursor, more = (
+        _get_augmented_skill_summaries_in_batches(
+            num_skills_to_fetch, urlsafe_start_cursor, sort_by))
+    filtered_augmented_skill_summaries = _filter_skills_by_status(
+        augmented_skill_summaries_batch, status)
+    filtered_augmented_skill_summaries = _filter_skills_by_classroom(
+        filtered_augmented_skill_summaries, classroom_name)
+
+    filtered_augmented_skill_summaries = _filter_skills_by_keywords(
+        filtered_augmented_skill_summaries, keywords)
+    augmented_skill_summaries.extend(filtered_augmented_skill_summaries)
+
+    while (len(augmented_skill_summaries) < num_skills_to_fetch
```

Comment:
Can't you have the above section inside this while loop? Why break the first iteration out of the loop?

Reply:
Yeah you are right, Done